pkg/router/template/router.go

```diff
@@ -386,18 +386,27 @@ func (r *templateRouter) readState() error {
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
 func (r *templateRouter) Commit() {
-	r.lock.Lock()
-	defer r.lock.Unlock()
 
-	if !r.synced {
-		glog.V(4).Infof("Router state synchronized for the first time")
-		r.synced = true
-		r.stateChanged = true
-	}
+	needCommit := false
+
+	func() {
```

Comment:
Why try so hard for defer?  I would simply reorganize this with an inline lock.  Defer makes it harder to reason about in this case.